system/extensions/esp32/wifi.toit

```diff
@@ -118,6 +137,18 @@ class WifiServiceDefinition extends NetworkServiceDefinitionBase:
 
   rssi resource/NetworkResource -> int?:
     return (state_.module as WifiModule).rssi
+  
+  scan config/Map -> List?:
+    if module_:
+      if module_.ap:
+        throw "wifi already established in AP mode"
+    else:
+      module_ = WifiModule.sta this "" ""
+
+    channel := config.get wifi.CONFIG_SCAN_CHANNEL
+    passive := config.get wifi.CONFIG_SCAN_PASSIVE
+    period  := config.get wifi.CONFIG_SCAN_PERIOD
+    return module_.scan channel passive period
```

Comment:
It feels like this is the problem with the `module` handling. You want to support multiple calls to scan without calling the `wifi_init` primitive multiple times? If that isn't necessary then I would call `state_.up` and `state_.down` from in here, but probably check that `state_.module` is null before doing that if scanning cannot work while we're connected.

```suggestion
    if state_.module:
      throw "wifi already connected or scanning in process"
    module := (state_.up: WifiModule.sta this "" "") as WifiModule
    try:
      channel := config.get wifi.CONFIG_SCAN_CHANNEL
      passive := config.get wifi.CONFIG_SCAN_PASSIVE
      period  := config.get wifi.CONFIG_SCAN_PERIOD
      return module.scan channel passive period
    finally:
      state_.down
```

Reply:
Yes, I would like to scan multi times without repeating to initialize Wi-Fi, because it costs much time. Here using `state_.up` will trigger a connection with ssid="" and password="", so this is not what I expect. What I would like to is that deinitialize Wi-Fi and free resource owned by scan when somewhere calls "connect" or "establish".

Reply:
In that case, I think we should move towards a model where we should go for a init/use*/close model like the network in AP and STA mode. This probably means that we should be returning a resource proxy for the network in the 'ready-to-scan' state and react with `state_.down` when it closes.

Do you need help figuring out how to do that?

Reply:
Yes, I want to know how to implement this.

Reply:
A good starting point might be the "Service resources and proxies" section in https://github.com/toitlang/toit/discussions/869. We would need to define a new subclass of `ServiceResource` that represents the network in the scanning state and return a handle for that resource through the `wifi.scanning` API call. That handle can then be turned into a subclass of `ServiceResourceProxy` on the client side and this is where you would add the `scan` method. The `scan` method will pass the handle back through the API and the WiFi service will lookup up the resources (using the handle as key) and act on that.

Reply:
I study this, but it is much complex to adding scan function by this method, and event it beyond to what I am able do ... so I decrease some performance to the scan, just one API that scan a set of channels, I think this can cover much use case, because in most applications users may scan all channels and determine to connect which one, if they don't find AP what need, they will also scan all channels, the time cost of initializing Wi-Fi is much less than scan all channels.

In my first opinion, I try to add APIs to scan one channel, and then one by one in sequence, so I have to keep the Wi-Fi is open, but if scan a set of channels one time, we may not need to keep the Wi-Fi open.